src/mongocrypt-kms-ctx.c

```diff
@@ -1378,4 +1479,146 @@ _mongocrypt_kms_ctx_init_gcp_decrypt (mongocrypt_kms_ctx_t *kms,
+
+   if (!kms_ctx->req) {
+      CLIENT_ERR ("Error creating KMIP register request: %s",
+                  kms_request_get_error (kms_ctx->req));
+      goto fail;
+   }
+
+   _mongocrypt_buffer_init (&kms_ctx->msg);
+   kms_ctx->msg.data = (uint8_t *) kms_request_to_bytes (kms_ctx->req, &reqlen);
+   if (!size_to_uint32 (reqlen, &kms_ctx->msg.len)) {
+      CLIENT_ERR ("Error casting request length %zu to uint32_t", reqlen);
+      goto fail;
+   }
+   kms_ctx->msg.owned = false;
+
+   ret = true;
+fail:
+   return ret;
+}
+
+bool
+_mongocrypt_kms_ctx_init_kmip_activate (mongocrypt_kms_ctx_t *kms_ctx,
+                                        const _mongocrypt_endpoint_t *endpoint,
+                                        const char *unique_identifier,
+                                        _mongocrypt_log_t *log)
+{
+   mongocrypt_status_t *status;
+   bool ret = false;
+   size_t reqlen;
+
+   _init_common (kms_ctx, log, MONGOCRYPT_KMS_KMIP_ACTIVATE);
+   status = kms_ctx->status;
+
+   kms_ctx->endpoint = bson_strdup (endpoint->host_and_port);
+   kms_ctx->req =
+      kms_kmip_request_activate_new (NULL /* reserved */, unique_identifier);
+
+   if (!kms_ctx->req) {
+      CLIENT_ERR ("Error creating KMIP activate request: %s",
+                  kms_request_get_error (kms_ctx->req));
+      goto fail;
+   }
+
+   _mongocrypt_buffer_init (&kms_ctx->msg);
+   kms_ctx->msg.data = (uint8_t *) kms_request_to_bytes (kms_ctx->req, &reqlen);
+   if (!size_to_uint32 (reqlen, &kms_ctx->msg.len)) {
+      CLIENT_ERR ("Error casting request length %zu to uint32_t", reqlen);
+      goto fail;
+   }
+   kms_ctx->msg.owned = false;
+
+   ret = true;
+fail:
+   return ret;
+}
+
+bool
+_mongocrypt_kms_ctx_init_kmip_get (mongocrypt_kms_ctx_t *kms_ctx,
+                                   const _mongocrypt_endpoint_t *endpoint,
+                                   const char *unique_identifier,
+                                   _mongocrypt_log_t *log)
+{
+   mongocrypt_status_t *status;
+   bool ret = false;
+   size_t reqlen;
+
+   _init_common (kms_ctx, log, MONGOCRYPT_KMS_KMIP_GET);
+   status = kms_ctx->status;
+
+   kms_ctx->endpoint = bson_strdup (endpoint->host_and_port);
+   kms_ctx->req =
+      kms_kmip_request_get_new (NULL /* reserved */, unique_identifier);
+
+   if (!kms_ctx->req) {
+      CLIENT_ERR ("Error creating KMIP get request: %s",
+                  kms_request_get_error (kms_ctx->req));
+      goto fail;
+   }
+
+   _mongocrypt_buffer_init (&kms_ctx->msg);
+   kms_ctx->msg.data = (uint8_t *) kms_request_to_bytes (kms_ctx->req, &reqlen);
+   if (!size_to_uint32 (reqlen, &kms_ctx->msg.len)) {
+      CLIENT_ERR ("Error casting request length %zu to uint32_t", reqlen);
+      goto fail;
+   }
+   kms_ctx->msg.owned = false;
+
+   ret = true;
+fail:
+   return ret;
+}
+
+const char *
+mongocrypt_kms_ctx_get_kms_provider (mongocrypt_kms_ctx_t *kms, uint32_t *len)
+{
+#define SET_AND_RET(x)                              \
```

Comment:
+1 (no need to adopt my perhaps-odd usage of the comma operator if it gives you the creeps (I like it because it "says" that the two operations MUST go together, but YYMV)). Thinking about it more, I'm also pretty confident that the optimizer will probably also handle this well if it's a helper function (it would be good to measure, but if this doesn't need every tic, I think we're ok either way).